File: cuecard-mobile/ios/CueCard/CueCard/Models/RemoteNotification.swift
import Foundation

/// The payload behind `GET /v2/notifications` on the mobile worker. Mirrors
/// `src/types.ts` there — change both together.
///
/// Decoding is deliberately forgiving. Anything a newer worker adds that this
/// build doesn't understand is skipped rather than taking the whole payload down
/// with it. The failure mode we want is always "nothing to show", never "app is
/// broken".
struct RemoteNotifications: Equatable {
    var notifications: [RemoteNotification]

    static let empty = RemoteNotifications(notifications: [])
}

/// One thing to show, on one surface.
struct RemoteNotification: Identifiable, Equatable {
    /// Where it's rendered. A surface this build doesn't know about is one it
    /// can't draw, so the notification is dropped in decoding.
    enum Surface: String, Decodable {
        case homeBanner
        case settingsRow
    }

    enum Severity: String, Decodable {
        case info
        case warning
        case critical
    }

    struct Action: Equatable {
        enum Kind: String, Decodable {
            case openURL
            case appStore
            case dismiss
        }

        var kind: Kind
        var label: String
        var url: URL?
    }

    var id: String
    var surface: Surface
    var severity: Severity = .info
    var priority = 0
    var title: String
    var body: String?
    var actions: [Action] = []
    var dismissible = true
    var expiresAt: Date?

    /// Hosts a notification is allowed to link to, mirrored from the worker. The
    /// worker no longer checks these itself, so this is the only enforcement:
    /// a link anywhere else means the notification is dropped rather than shown.
    static let allowedHosts: Set<String> = [
        "cuecard.dev", "www.cuecard.dev", "apps.apple.com", "github.com"
    ]

    /// Whether this is safe and sensible to draw: a title, no more than two
    /// actions, every link pointing somewhere we recognise, and some way out.
    var isRenderable: Bool {
        guard !title.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else { return false }
        guard actions.count <= 2 else { return false }

        for action in actions where action.kind == .openURL {
            guard let host = action.url?.host, Self.allowedHosts.contains(host) else { return false }
        }

        return dismissible || !actions.isEmpty
    }

    /// Expiry is enforced here alone — the worker serves the same list to
    /// everyone and doesn't filter on it.
    func hasExpired(asOf now: Date = Date()) -> Bool {
        guard let expiresAt else { return false }
        return expiresAt <= now
    }
}

// MARK: - Decoding

/// Decodes an element, or nothing at all. Lets one malformed or not-yet-understood
/// notification drop out without discarding the ones around it.
private struct Skipping<Wrapped: Decodable>: Decodable {
    let value: Wrapped?

    init(from decoder: Decoder) throws {
        value = try? Wrapped(from: decoder)
    }
}

/// The worker writes plain ISO 8601; accept fractional seconds too in case that
/// ever changes underneath us.
private func parseISO8601(_ string: String) -> Date? {
    let withFraction = ISO8601DateFormatter()
    withFraction.formatOptions = [.withInternetDateTime, .withFractionalSeconds]

    return ISO8601DateFormatter().date(from: string) ?? withFraction.date(from: string)
}

extension RemoteNotifications: Decodable {
    private enum CodingKeys: String, CodingKey {
        case notifications
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)

        let decoded = (try? container.decode([Skipping<RemoteNotification>].self, forKey: .notifications)) ?? []
        notifications = decoded.compactMap(\.value)
    }
}

extension RemoteNotification: Decodable {
    private enum CodingKeys: String, CodingKey {
        case id, surface, severity, priority, title, body
        case actions, dismissible, expiresAt
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)

        // An id, a surface and a title are the notification. Without all three
        // there's nothing to draw or remember, so let the decode fail and skip it.
        id = try container.decode(String.self, forKey: .id)
        surface = try container.decode(Surface.self, forKey: .surface)
        title = try container.decode(String.self, forKey: .title)

        severity = (try? container.decode(Severity.self, forKey: .severity)) ?? .info
        priority = (try? container.decode(Int.self, forKey: .priority)) ?? 0
        body = try? container.decode(String.self, forKey: .body)
        dismissible = (try? container.decode(Bool.self, forKey: .dismissible)) ?? true

        if let expiry = try? container.decode(String.self, forKey: .expiresAt) {
            expiresAt = parseISO8601(expiry)
        }

        // An action kind we don't recognise is one we can't carry out, and the
        // action is usually the point — so drop the notification rather than show
        // a card whose button does nothing.
        actions = try container.decodeIfPresent([Action].self, forKey: .actions) ?? []
    }
}

extension RemoteNotification.Action: Decodable {
    private enum CodingKeys: String, CodingKey {
        case kind, label, url
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        kind = try container.decode(Kind.self, forKey: .kind)
        label = try container.decode(String.self, forKey: .label)
        url = try? container.decode(URL.self, forKey: .url)
    }
}

File: cuecard-mobile/android/app/src/main/java/com/thisisnsh/cuecard/android/Models/RemoteNotification.kt
package com.thisisnsh.cuecard.android.models

import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.jsonObject
import java.net.URI
import java.time.Instant
import java.time.OffsetDateTime
import java.time.format.DateTimeFormatter

/**
 * The payload behind `GET /v2/notifications` on the mobile worker. Mirrors
 * `src/types.ts` there — change both together.
 *
 * Decoding is deliberately forgiving. Anything a newer worker adds that this
 * build doesn't understand is skipped rather than taking the whole payload down
 * with it. The failure mode we want is always "nothing to show", never "app is
 * broken".
 */
data class RemoteNotifications(
    val notifications: List<RemoteNotification> = emptyList()
) {
    companion object {
        val EMPTY = RemoteNotifications()

        /** Lenient by design: see the class comment. */
        private val json = Json {
            ignoreUnknownKeys = true
            isLenient = true
        }

        /**
         * Decode a worker response. One unreadable notification drops out without
         * taking the rest of the payload with it — the parallel to the iOS
         * `Skipping<Wrapped>` wrapper.
         */
        fun decode(raw: String): RemoteNotifications? {
            val root = runCatching { json.parseToJsonElement(raw).jsonObject }.getOrNull() ?: return null

            val notifications = (root["notifications"] as? JsonArray).orEmpty().mapNotNull { element ->
                runCatching { json.decodeFromJsonElement(RemoteNotificationDTO.serializer(), element) }
                    .getOrNull()
                    ?.toNotification()
            }

            return RemoteNotifications(notifications = notifications)
        }
    }
}

/** One thing to show, on one surface. */
data class RemoteNotification(
    val id: String,
    val surface: Surface,
    val severity: Severity = Severity.INFO,
    val priority: Int = 0,
    val title: String,
    val body: String? = null,
    val actions: List<Action> = emptyList(),
    val dismissible: Boolean = true,
    val expiresAt: Instant? = null
) {
    /**
     * Where it's rendered. A surface this build doesn't know about is one it
     * can't draw, so the notification is dropped in decoding.
     */
    enum class Surface(val rawValue: String) {
        HOME_BANNER("homeBanner"),
        SETTINGS_ROW("settingsRow");

        companion object {
            fun from(value: String?): Surface? = entries.find { it.rawValue == value }
        }
    }

    enum class Severity(val rawValue: String) {
        INFO("info"),
        WARNING("warning"),
        CRITICAL("critical");

        companion object {
            fun from(value: String?): Severity? = entries.find { it.rawValue == value }
        }
    }

    data class Action(
        val kind: Kind,
        val label: String,
        val url: String? = null
    ) {
        enum class Kind(val rawValue: String) {
            OPEN_URL("openURL"),
            APP_STORE("appStore"),
            DISMISS("dismiss");

            companion object {
                fun from(value: String?): Kind? = entries.find { it.rawValue == value }
            }
        }
    }

    /**
     * Whether this is safe and sensible to draw: a title, no more than two
     * actions, every link pointing somewhere we recognise, and some way out.
     */
    val isRenderable: Boolean
        get() {
            if (title.trim().isEmpty()) return false
            if (actions.size > 2) return false

            for (action in actions.filter { it.kind == Action.Kind.OPEN_URL }) {
                val url = action.url ?: return false
                val host = runCatching { URI(url).host }.getOrNull() ?: return false
                if (!allowedHosts.contains(host)) return false
            }

            return dismissible || actions.isNotEmpty()
        }

    /**
     * Expiry is enforced here alone — the worker serves the same list to everyone
     * and doesn't filter on it.
     */
    fun hasExpired(now: Instant = Instant.now()): Boolean {
        val expiry = expiresAt ?: return false
        return !expiry.isAfter(now)
    }

    companion object {
        /**
         * Hosts a notification is allowed to link to. The worker no longer checks
         * these itself, so this is the only enforcement: a link anywhere else
         * means the notification is dropped rather than shown without its action.
         */
        val allowedHosts = setOf(
            "cuecard.dev", "www.cuecard.dev", "play.google.com", "github.com"
        )
    }
}

// MARK: - Decoding

/**
 * The worker writes plain ISO 8601; accept fractional seconds too in case that
 * ever changes underneath us.
 */
private fun parseISO8601(value: String): Instant? =
    runCatching { OffsetDateTime.parse(value, DateTimeFormatter.ISO_OFFSET_DATE_TIME).toInstant() }
        .recoverCatching { Instant.parse(value) }
        .getOrNull()

@Serializable
private data class ActionDTO(
    val kind: String,
    val label: String,
    val url: String? = null
) {
    /**
     * An action kind we don't recognise is one we can't carry out, and the action
     * is usually the point — so this throws, and the notification goes with it.
     */
    fun toAction(): RemoteNotification.Action {
        val parsed = RemoteNotification.Action.Kind.from(kind)
            ?: throw IllegalArgumentException("Unknown action kind: $kind")
        return RemoteNotification.Action(kind = parsed, label = label, url = url)
    }
}

@Serializable
private data class RemoteNotificationDTO(
    val id: String,
    val surface: String,
    val severity: String? = null,
    val priority: Int? = null,
    val title: String,
    val body: String? = null,
    val actions: List<ActionDTO>? = null,
    val dismissible: Boolean? = null,
    val expiresAt: String? = null
) {
    /**
     * An id, a surface and a title are the notification. Without all three there's
     * nothing to draw or remember, so this throws and it is skipped.
     */
    fun toNotification(): RemoteNotification {
        val parsedSurface = RemoteNotification.Surface.from(surface)
            ?: throw IllegalArgumentException("Unknown surface: $surface")

        return RemoteNotification(
            id = id,
            surface = parsedSurface,
            severity = RemoteNotification.Severity.from(severity) ?: RemoteNotification.Severity.INFO,
            priority = priority ?: 0,
            title = title,
            body = body,
            actions = actions?.map { it.toAction() } ?: emptyList(),
            dismissible = dismissible ?: true,
            expiresAt = expiresAt?.let { parseISO8601(it) }
        )
    }
}
